include/swift/AST/ExtInfo.h

```diff
@@ -64,18 +70,54 @@ class ClangTypeInfo {
   //    clang::ReferenceType which points to a clang::FunctionType.
   //
   // When used as a part of SILFunctionType, the type is canonical.
-  const clang::Type *type;
+  //
+  // States:
+  // {ptr != 0, bool == true}  -> has Clang type, derivable from Swift type
+  // {ptr != 0, bool == false} -> has Clang type, not derivable from Swift type
+  // {ptr == 0, bool == true}  -> impossible
+  // {ptr == 0, bool == false} -> no Clang type, no derivability information
+  llvm::PointerIntPair<const clang::Type *, 1, bool> storage;
 
-  constexpr ClangTypeInfo() : type(nullptr) {}
-  constexpr ClangTypeInfo(const clang::Type *type) : type(type) {}
+  ClangTypeInfo(const clang::Type *type, bool b) : storage(type, b) {}
 
   friend bool operator==(ClangTypeInfo lhs, ClangTypeInfo rhs);
   ClangTypeInfo getCanonical() const;
 
 public:
-  constexpr const clang::Type *getType() const { return type; }
+  ClangTypeInfo() : storage(nullptr, false) {}
+
+  static ClangTypeInfo derivableFromSwiftType(const clang::Type *ty) {
+    assert(ty && "Use ClangTypeInfo() instead.");
+    return ClangTypeInfo(ty, true);
+  }
+
+  static ClangTypeInfo notDerivableFromSwiftType(const clang::Type *ty) {
+    assert(ty && "Use ClangTypeInfo() instead.");
+    return ClangTypeInfo(ty, false);
+  }
+
+  /// Check if we should prefer the parsed type for storage in an ExtInfo.
+  ///
+  /// Prefer the parsed type over the computed type in case they are not
+  /// exactly equivalent, since the parsed type will have greater fidelity
+  /// (sugar and attributes).
+  ///
+  /// NOTE: Sometimes, two canonical Clang types may differ in a way that does
+  /// not matter from a Swift perspective. For example, the parsed type may have
+  /// a "__kindof", which is never part of a computed type. We don't strip such
+  /// attributes. Instead, we return the parsed type marked non-derivable, even
+  /// if it is derivable upto differences of "__kindof".
```

Comment:
"up to"

The method name is odd here, since presumably we always want to use the parsed type but just need to set the is-canonical bit appropriately.

Reply:
Sometimes there is no parsed type (the argument is allowed to be `nullptr`). I don't like the method name much myself, I've been thinking about alternatives but haven't found something satisfying.